src/components/InngestFunction.ts

```diff
@@ -485,6 +499,30 @@ export class InngestFunction<
         state.nonStepFnDetected = true;
 
         return ["complete", await userFnPromise];
+      } else {
+        /**
+         * If we're here, the user's function has not returned a value, has not
+         * reported any new, ops, but has also previously used step tools and
+         * successfully memoized state.
+         *
+         * This indicates that the user has mixed step and non-step logic, which
+         * is not something we want to support without an opt-in from the user.
+         *
+         * We should throw here to let the user know that this is not supported.
+         *
+         * We need to be careful, though; it's a valid state for a chain of
+         * promises to return no further actions, so we should only throw if
+         * this state is reached and there are no other pending steps.
+         */
+        const hasOpsPending = Object.values(state.allFoundOps).some((op) => {
+          return op.fulfilled === false;
+        });
+
+        if (!hasOpsPending) {
+          throw new Error(
+            functionStoppedRunningErr(ErrCode.ASYNC_DETECTED_AFTER_MEMOIZATION)
```

Comment:
Should throwing any of these errors disable retries? e.g. return `4xx`. I think yes.

Reply:
I agree retries should not attempt since there's no point until user update their code